### PR TITLE
Fixed #19259 -- Added group by selected primary keys support.

### PR DESCRIPTION
**Pull request by charettes**
- [x] Enable the `allows_group_by_selected_pks` feature flag only for PostgreSQL 9.1+
- [x] Avoid unnecessary group by foreign keys

### CODE REVIEW

**Comment by charettes** [comment]
Builbot, test this please.

**Comment by charettes** [comment]
I'm trying to figure out what else might be failing here.

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -150,7 +147,15 @@ def collapse_group_by(self, expressions, having):
                         getattr(expr.output_field, 'model') == self.query.model):
```
Comment (charettes) [inline django/db/models/sql/compiler.py]: I think this approach will add an extraneous expression when a self referential foreign key is involved.

```python
class Person(models.Model):
    best_friend = models.ForeignKey('self', related_name='best_friends')
    age =  models.SmallPositiveIntegerField()

Person.objects.annotate(Avg('best_friends__age'))
```

File: django/db/backends/postgresql_psycopg2/features.py [hunk 2]
```diff
@@ -27,3 +28,7 @@ class DatabaseFeatures(BaseDatabaseFeatures):
     closed_cursor_error_class = InterfaceError
     has_case_insensitive_like = False
     requires_sqlparse_for_splitting = False
+
+    @cached_property
+    def allows_group_by_selected_pks(self):
+        return self.connection.pg_version >= 90100
```
Comment (charettes) [inline django/db/backends/postgresql_psycopg2/features.py]: This might not be required if we drop support for 9.0 in 1.9. @shaib pointed out that it will EOL in September.

File: django/db/models/sql/compiler.py (hunk 1 above)
Comment (charettes) [inline django/db/models/sql/compiler.py]: More like `Person.objects.select_related('best_friend').annotate(Avg('best_friends__age'))` actually.

File: django/db/models/sql/compiler.py [hunk 3]
```diff
@@ -150,7 +147,14 @@ def collapse_group_by(self, expressions, having):
...
                     pk = expr
```
Comment (jarshwah) [inline django/db/models/sql/compiler.py]: I know this isn't part of your code, but can we break out of the loop if this condition is met?

File: django/db/models/sql/compiler.py [hunk 4]
```diff
@@ -150,7 +147,14 @@ def collapse_group_by(self, expressions, having):
...
             if pk:
+                # MySQLism: Columns in HAVING clause must be added to the GROUP BY.
```
Comment (jarshwah) [inline django/db/models/sql/compiler.py]: It's good that you've documented that this condition is a mysql-ism, but I'm a bit suspect about including behaviour specific to mysql in a block that is gated by a feature that (theoretically) could apply to other backends. I guess we can deal with it if it becomes a problem (which this comment will help to find).

File: django/db/models/sql/compiler.py [hunk 5]
```diff
@@ -150,7 +147,14 @@ def collapse_group_by(self, expressions, having):
...
                 expressions = [pk] + [expr for expr in expressions if expr in having]
+        elif self.connection.features.allows_group_by_selected_pks:
+            pks = {expr for expr in expressions if hasattr(expr, 'target') and expr.target.primary_key}
+            aliases = {expr.alias for expr in pks}
+            expressions = [
+                expr for expr in expressions if expr in pks or getattr(expr, 'alias', None) not in aliases
```
Comment (jarshwah) [inline django/db/models/sql/compiler.py]: Include expression if it is a PK or if the alias is not one an alias of a PK? Wouldn't this include all fields that have an alias? (I know I must be reading this incorrectly because the tests pass..)

File: django/db/models/sql/compiler.py (hunk 3 above)
Comment (charettes) [inline django/db/models/sql/compiler.py]: I could make this change in another commit.

File: django/db/models/sql/compiler.py (hunk 4 above)
Comment (charettes) [inline django/db/models/sql/compiler.py]: I couldn't agree more but this is no invention of mine. I simply moved this comment down from the line 140.

File: django/db/models/sql/compiler.py (hunk 5 above)
Comment (charettes) [inline django/db/models/sql/compiler.py]: This actually collapse all expressions that have an alias already represented by a primary key in the provided expressions.

Maybe `expr in pks or not hasattr(expr, 'alias') or expr.alias not in aliases` would make it more readable?

File: django/db/models/sql/compiler.py (hunk 4 above)
Comment (akaariai) [inline django/db/models/sql/compiler.py]: The reason its here is that it has been implemented more or less this way for a long time (though the code hasn't been even this clear before compiler expressions refactor).

MySQL has been the only backend to reach this part of code. If some backend would allow grouping by primary key only, and wouldn't need expressions in having either, then that backend would get extra expressions in GROUP BY. This is non-optimal, but not an error.

Of course, if somebody wants to clean this up, +1 to that.

File: django/db/models/sql/compiler.py (hunk 5 above)
Comment (akaariai) [inline django/db/models/sql/compiler.py]: We want a comment here what we are doing, and why we are doing it.

I'm wondering what kind of expressions that do not have alias end up in here. Everything should come through get_group_by_cols(), and those cols should be "raw columns", that is direct references to alias.col. Seems like that isn't the case. It might be possible to fix this in a way that only alias.col references end up in here in the first place.

Comment (charettes) [inline django/db/models/sql/compiler.py]: From a quick investigation (running `aggregation_regress` with `expr.alias not in aliases`) the following expression with no alias end up here:

1. `CombinedExpression`
2. `RawSQL`
3. `Date`

> It might be possible to fix this in a way that only alias.col references end up in here in the first place.

Should we tackle it in this patch?

Comment (akaariai) [inline django/db/models/sql/compiler.py]: RawSQL needs to be added to the group by clause, we can't know if it refers to something that doesn't have an existing alias.

Getting CombinedExpression and Date here seems curious. These should be fixable, but lets not do it in this patch.

Comment (charettes) [inline django/db/models/sql/compiler.py]: Thanks for the follow up. I added a comment about the functional dependency reduction logic.

**Comment by charettes** [comment]
Buildbot, test this please.

File: django/db/backends/postgresql_psycopg2/features.py (hunk 2 above)
Comment (timgraham) [inline django/db/backends/postgresql_psycopg2/features.py]: Good for me: https://github.com/django/django/pull/4399.

File: django/db/models/sql/compiler.py [hunk 6]
```diff
@@ -150,7 +147,17 @@ def collapse_group_by(self, expressions, having):
                         getattr(expr.output_field, 'model') == self.query.model):
                     pk = expr
             if pk:
+                # MySQLism: Columns in HAVING clause must be added to the GROUP BY.
                 expressions = [pk] + [expr for expr in expressions if expr in having]
+        elif self.connection.features.allows_group_by_selected_pks:
+            # Filter out all expressions associated with a table's primary key present in the grouped columns.
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: We usually wrap comments at 79 chars, LGTM otherwise if the ORM guys are happy with it.

**Comment by charettes** [comment]
Wrapped the comments at 79 chars. Will merge once I get final approval from the ORM guys.

**Comment by jarshwah** [comment]
LGTM :shipit:
